src/main/kotlin/com/manimdsl/ManimParserVisitor.kt

```diff
@@ -7,9 +7,66 @@ import com.manimdsl.frontend.*
 class ManimParserVisitor : ManimParserBaseVisitor<ASTNode>() {
     val symbolTable = SymbolTableVisitor()
     private val semanticAnalyser = SemanticAnalysis()
+    private var inFunction: Boolean = false
+    private var functionReturnType: Type = VoidType
 
```

Comment:
If you can add functions to the symbol table when they are called if they haven't yet been declared then you can verify their type later on by looking it up in the symbol table when the function is declared rather than when it is called. It just means an "inferred" flag or something needs to be added to the function type so that redeclarations of functions remain illegal.

In the visiting of the top level ProgramNode after constructing the AST you can then conduct a "undefinedFunctionCall" check after constructing the AST by checking if any function in the symbol table is actually in the AST if you add functions at the call point to the symbol table but if they were never declared then they won't be in the list of functions of the AST ProgramNode.

Happy to help out with this if it's confusing!

Reply:
This was for the forward declarations thing you mentioned

Reply:
Does this mean we will add the semantic error when we visit the function declaration (if the type doesn't match)? And how do we handle multiple assignments with function calls to the same function (defined later) with different type on the lhs? (how do I know which type should be the compatible one or if both are incompatible?)

Reply:
Yes sure. Good point the error messages might not be great but we can just throw an error when the same function is called in different ways and say something like "function f called in 2 different ways". I'm sure there's a way but this would do for now I'm sure we can revisit the error messages anyway later.